### PR TITLE
Refactor make a code more generic - Closes #38

### LINKED ISSUE #38: Make a code more generic and use object oriented programming
### Description

Make a code more generic by using traits and implementing them on structs or types.

- currently some functions like `key_hash()`, `value_hash()`, etc. are used only on `&[u8]` object type
- we may introduce a trait for those functions and implement it for an object type
- example: instead of a function call `key_hash(&value)`, we could call function `value.key_hash()` on an object
- reference implementation:

```
pub trait MyHash {
  fn my_hash(&self) -> Vec<u8>;
}

impl MyHash for &[u8] {
  fn my_hash(&self) -> Vec<u8> {
	let mut hasher = Sha256::new();
	hasher.update(self);
	let result = hasher.finalize();
	return result.as_slice().to_vec();
  }
}
```

### Acceptance Criteria

- traits are used in a code
- methods are implemented for an objects where it makes sense

### Additional Information

- only perform this action on a places in the code where and if it makes sense

### PR DESCRIPTION
**Pull request by hrmhatef**
### What was the problem?

This PR resolves #38 

### How was it solved?

 - Introduce new traits and implement them for types that needs.

### How was it tested?

All tests passed.

### CODE REVIEW

**Review by matjazv — CHANGES_REQUESTED** [review]
Very good!

File: src/state_writer.rs [hunk 1]
```diff
@@ -1,17 +1,20 @@
 use neon::prelude::*;
 use neon::types::buffer::TypedArray;
-use std::cell::RefCell;
+// use std::cell::RefCell;
```
[the remark this answers is not in the document]
Comment (hrmhatef) [reply src/state_writer.rs]: Done.

**Review by matjazv — APPROVED** [review]

File: src/smt.rs [hunk 2]
```diff
@@ -149,43 +149,15 @@ pub struct SparseMerkleTree {
...
-    hasher.update(node_hash);
-    let result = hasher.finalize();
-    result.to_vec()
-}
-
-fn empty_hash() -> Vec<u8> {
-    let hasher = Sha256::new();
-    let result = hasher.finalize();
-    result.to_vec()
+impl Hash256 for KVPair {
```
Comment (shuse2) [inline src/smt.rs:152]: this Hash256 implementation is quite specific to the SMT implementation while KVPair is quite generic type?

**Review by shuse2 — APPROVED** [review]
LGTM!
Maybe one things to consider further in the future is the domain separation of each module / file? especially `common_db` and `types`
From the module name, those are bit hard to tell the contents, so maybe it's better to consider separation/placement of those logics again

File: src/smt.rs (hunk 2 above)
Comment (hrmhatef) [reply src/smt.rs:152]: Yes, We need a specific implementation of the `Hash256` for `KVPair` to use it in `SMT`.

**Comment by hrmhatef** [comment]
> LGTM! Maybe one things to consider further in the future is the domain separation of each module / file? especially `common_db` and `types` From the module name, those are bit hard to tell the contents, so maybe it's better to consider separation/placement of those logics again

I agree with you, we can do that in issue #50, I can add a subtask for this on that issue. Is it Okay?

File: src/smt.rs (hunk 2 above)
Comment (shuse2) [reply src/smt.rs:152]: I meant `KVPair` is quite generic type while it has specific SMT implementation seemed bit strange but it's under `smt` module so very minor point 👍

**Pull request merged**